### PR TITLE
[REF] point_of_sale, pos_*: adapt generic components to PoS context.

### PR DESCRIPTION
**Pull request by davidmonnom**
*: l10n_co_pos, l10n_es_edi_tbai_pos, l10n_es_pos, l10n_fr_pos_cert,
l10n_gcc_pos, l10n_in_pos, l10n_sa_pos, pos_loyalty, pos_online_payment,
pos_online_payment_self_order, pos_restaurant, pos_sale, pos_self_order

Components in the Point of Sale (PoS) context should always be able to
receive related model objects as props.

Usually, if a component is only used in PoS or Self Order, it can use
services to get the needed object. However, if a component is used in
both Self Order and PoS, it cannot use services to get the object
because the service will be different in each context.

Since Self Order and PoS both use the same models, we can pass the model
object as props to the component.

PoS context components should not be used in the customer display. Since
the customer display is a separate context, it receives pre-processed
data from PoS. Therefore, we need to create a new component for the
customer display.

### CODE REVIEW

File: addons/l10n_gcc_pos/static/src/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 1]
```diff
@@ -14,61 +14,61 @@
...
             </div>
         </xpath>
 
-        <xpath expr="//t[@t-if='props.data.rounding_applied']" position="attributes">
-            <attribute name="t-if" add=" and !props.data.is_gcc_country" separator=" "/>
-        </xpath>
-        <xpath expr="//t[@t-if='props.data.rounding_applied and !props.data.is_gcc_country']" position="after">
-            <t t-elif="props.data.rounding_applied and props.data.is_gcc_country">
+        <xpath expr="//t[@t-if='roundAmount']" position="after">
+            <t t-elif="order.getRoundingApplied() and order.isGccCountry">
```
Comment (caburj) [inline addons/l10n_gcc_pos/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:39]:
```suggestion
            <t t-elif="order.getRoundingApplied() and !order.isGccCountry">
```

File: addons/l10n_gcc_pos/static/src/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 2]
```diff
@@ -14,61 +14,61 @@
...
                 <div class="pos-receipt-amount">
                     To Pay
-                    <span t-esc="props.formatCurrency(props.data.amount_total + props.data.rounding_applied)" class="pos-receipt-right-align"/>
+                    <span t-esc="formatCurrency(order.getTotalWithTax() + order.getRoundingApplied())" class="pos-receipt-right-align"/>
                 </div>
             </t>
         </xpath>
+        <xpath expr="//t[@t-if='roundAmount']" position="attributes">
+            <attribute name="t-if" add=" and !order.isGccCountry" separator=" "/>
+        </xpath>
```
Comment (caburj) [inline addons/l10n_gcc_pos/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:52]: perhaps you don't need this if you apply the recommendation [above](https://github.com/odoo/odoo/pull/189446/files#r1873002808).

File: addons/point_of_sale/static/src/app/components/order_display/order_display.xml [hunk 3]
```diff
@@ -1,50 +1,52 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <templates id="template" xml:space="preserve">
-    <t t-name="point_of_sale.OrderWidget">
-        <t t-if="props.lines?.length">
+    <t t-name="point_of_sale.OrderDisplay">
+        <t t-if="order.lines.length">
+            <div t-attf-class="{{console.log(order.lines)}}" />
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/components/order_display/order_display.xml:5]: should remove console.log

File: addons/point_of_sale/static/src/app/components/order_display/order_display.xml [hunk 4]
```diff
@@ -1,50 +1,52 @@
...
                             <li t-if="note.trim() !== ''" t-esc="note"
                                 class="internal-note badge mt-1 p-2 rounded-pill bg-info text-info bg-opacity-25"
                                 style="font-size: 0.85rem;"/>
                         </t>
                     </div>
                 </div>
-                <div t-if="props.total or props.tax" class="order-summary d-flex flex-column gap-1 px-3 py-2 border-top fw-bolder lh-sm">
-                    <div t-if="props.tax" class="tax-info subentry d-flex justify-content-between w-100 fs-6 text-muted ">
+                <t t-set="orderTotal" t-value="formatCurrency(order.getTotalWithTax())"/>
+                <t t-set="orderTax" t-value="!floatIsZero(order.getTotalTax()) and formatCurrency(order.getTotalTax()) or ''"/>
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/components/order_display/order_display.xml:39]:
```suggestion
                <t t-set="orderTax" t-value="!floatIsZero(order.getTotalTax()) ? formatCurrency(order.getTotalTax()) : ''"/>
```

File: addons/point_of_sale/static/src/app/components/order_display/order_display.xml [hunk 5]
```diff
@@ -1,50 +1,52 @@
...
+                <div t-if="(orderTotal or orderTax) and props.mode !== 'receipt'" class="order-summary d-flex flex-column gap-1 px-3 py-2 border-top fw-bolder lh-sm">
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/components/order_display/order_display.xml:40]:
```suggestion
                <div t-if="props.mode !== 'receipt'" class="order-summary d-flex flex-column gap-1 px-3 py-2 border-top fw-bolder lh-sm">
```

you can remove this condition, it's useless.

File: addons/point_of_sale/static/src/app/components/orderline/orderline.js [hunk 6]
```diff
@@ -1,36 +1,46 @@
...
+    }
+
+    get packLotLines() {
+        return this.line.pack_lot_ids.map(
+            (l) =>
+                `${l.pos_order_line_id.product_id.tracking == "lot" ? "Lot Number" : "SN"} ${
+                    l.lot_name
+                }`
+        );
+    }
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/components/orderline/orderline.js:45]: duplicate. use the getter from pos_order_line model.

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.js [hunk 7]
```diff
@@ -1,28 +1,63 @@
...
+    get qrCode() {
+        const baseUrl = this.order.session._base_url;
+        return (
+            this.order.company.point_of_sale_use_ticket_qr_code &&
+            this.order.finalized &&
+            qrCodeSrc(`${baseUrl}/pos/ticket/`)
+        );
+    }
+
+    get formatedDate() {
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.js:44]:
```suggestion
    get formattedShippingDate() {
```

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 8]
```diff
@@ -122,18 +123,18 @@
             </div>
 
             <br/>
-            <t t-if="props.data.shippingDate">
+            <t t-if="order.shipping_date">
                 <div class="pos-receipt-order-data">
                     Expected delivery:
-                    <div><t t-esc="props.data.shippingDate" /></div>
+                    <div><t t-esc="formatedDate" /></div>
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:129]:
```suggestion
                    <div><t t-esc="formattedShippingDate" /></div>
```

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 9]
```diff
@@ -3,115 +3,116 @@
     <t t-name="point_of_sale.OrderReceipt">
         <div class="pos-receipt p-2">
             <t t-set="showTaxGroupLabels" t-value="doesAnyOrderlineHaveTaxLabel()"/>
-            <ReceiptHeader data="props.data.headerData" />
-            <OrderWidget lines="props.data.orderlines" t-slot-scope="scope"
-                generalCustomerNote="props.data.general_customer_note or ''"
-                internalNote="props.data.internalNote or ''" screenName="props.data.screenName">
+            <ReceiptHeader data="this.header" order="order" />
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:6]:
```suggestion
            <ReceiptHeader order="order" />
```
Like you do with the receipt, compute the header data from the ReceiptHeader component using the given order.

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.js [hunk 10]
```diff
@@ -1,28 +1,63 @@
...
+        return formatDate(DateTime.fromSQL(this.order.shipping_date));
+    }
+
+    get orderDate() {
+        return formatDateTime(parseUTCString(this.order.date_order));
+    }
+
+    get paymentLine() {
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.js:52]:
```suggestion
    get paymentLines() {
```

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 11]
```diff
@@ -3,115 +3,116 @@
...
+           <div t-if="order.config.receipt_footer" class="pos-receipt-center-align" style="white-space:pre-line">
+                <br/>
+                <t t-esc="order.config.receipt_footer" />
                 <br/>
                 <br/>
             </div>
 
             <div class="after-footer">
-                <t t-foreach="props.data.paymentlines" t-as="line" t-key="line_index">
+                <t t-foreach="paymentLine" t-as="line" t-key="line_index">
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:115]:
```suggestion
                <t t-foreach="paymentLines" t-as="line" t-key="line_index">
```

File: addons/point_of_sale/static/src/app/services/pos_store.js [hunk 12]
```diff
@@ -1659,7 +1652,6 @@ export class PosStore extends WithLazyGetterTrap {
         }
         const payload = await makeAwaitable(this.dialog, PartnerList, {
             partner: currentPartner,
-            getPayload: (newPartner) => currentOrder.setPartner(newPartner),
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/services/pos_store.js:1662]: I don't think you should remove this line.

File: addons/pos_restaurant/static/src/app/screens/receipt_screen/order_receipt/order_receipt.xml [hunk 13]
```diff
@@ -2,33 +2,28 @@
 <templates id="template" xml:space="preserve">
 
     <t t-name="pos_restaurant.OrderReceipt" t-inherit="point_of_sale.OrderReceipt" t-inherit-mode="extension">
-        <xpath expr="//ReceiptHeader" position="after">
-            <div t-if="props.data.isBill" class="fs-2 mb-1">Pro forma receipt</div>
-        </xpath>
-        <xpath expr="//div[hasclass('receipt-change')]" position="attributes">
-            <attribute name="t-if">'order_change' in taxTotals and !props.data.isBill</attribute>
-        </xpath>
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/receipt_screen/order_receipt/order_receipt.xml:10]: is this moved somewhere else?

File: addons/l10n_gcc_pos/static/src/app/screens/receipt_screen/receipt/order_receipt.xml
Comment (caburj) [inline addons/l10n_gcc_pos/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:1]: looks like you made a wrong rebase here. You might need to remove the added code in this file.

File: addons/l10n_gcc_pos/static/src/app/models/pos_order.js [hunk 14]
```diff
@@ -1,19 +1,8 @@
...
-        const results = super.exportForPrinting(...arguments);
-        results.is_gcc_country = ["SA", "AE", "BH", "OM", "QA", "KW"].includes(
-            this.company.country_id?.code
-        );
-        if (results.is_gcc_country) {
-            results.label_total = _t("TOTAL / اﻹجمالي");
-            results.label_rounding = _t("Rounding / التقريب");
-            results.label_change = _t("CHANGE / الباقي");
-            results.label_discounts = _t("Discounts / الخصومات");
-        }
```
Comment (caburj) [inline addons/l10n_gcc_pos/static/src/app/models/pos_order.js:16]: don't forget about these labels. I believe they're introduced recently.

File: addons/pos_restaurant/static/src/app/screens/receipt_screen/order_receipt/order_receipt.xml (hunk 13 above)
Comment (davidmonnom) [reply addons/pos_restaurant/static/src/app/screens/receipt_screen/order_receipt/order_receipt.xml]: I add it back, now is a proforma if the order isn't finalized

File: addons/point_of_sale/static/src/app/components/order_widget/order_widget.xml [hunk 15]
```diff
@@ -1,75 +0,0 @@
...
-                    </div>
-                    <div t-if="'cash_rounding_base_amount_currency' in taxTotals"
-                         class="subentry d-flex justify-content-between w-100 fs-6 text-muted"
-                    >
-                        Rounding
-                        <div id="order-widget-rounding">
-                            <span t-esc="formatMonetary(taxTotals.cash_rounding_base_amount_currency, {currencyId: taxTotals.currency_id})"
-                                  class="rounding"/>
-                        </div>
-                    </div>
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/components/order_widget/order_widget.xml:57]: Looks like the new `addons/point_of_sale/static/src/app/components/order_display/order_display.xml` file doesn't take into account these lines.

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 16]
```diff
@@ -1,122 +1,122 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <templates id="template" xml:space="preserve">
     <t t-name="point_of_sale.OrderReceipt">
+    <!-- ults.is_gcc_country) {
+            results.label_total = _t("TOTAL / اﻹجمالي");
+            results.label_rounding = _t("Rounding / التقريب");
+            results.label_change = _t("CHANGE / الباقي");
+            results.label_discounts = _t("Discounts / الخصومات"); -->
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:8]: ?

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 17]
```diff
@@ -1,122 +1,122 @@
...
-                <Orderline basic_receipt="props.basic_receipt" line="omit(scope.line, 'customerNote')" class="{ 'px-0': true }" showTaxGroupLabels="showTaxGroupLabels">
+                <Orderline line="line" showTaxGroup="showTaxGroupLabels" mode="'receipt'">
                     <li t-if="line.customerNote" class="customer-note w-100 p-2 my-1 rounded text-break">
                         <i class="fa fa-sticky-note me-1" role="img" aria-label="Customer Note" title="Customer Note"/>
                         <t t-esc="line.customerNote" />
                     </li>
                 </Orderline>
-            </OrderWidget>
-            <t t-if="!props.basic_receipt">
-                <t t-set="taxTotals" t-value="props.data.taxTotals"/>
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:19]: The order_receipt component is now using `taxTotals`, not `getTaxDetails()`. Adapt correctly.

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml (hunk 16 above)
Comment (davidmonnom) [reply addons/point_of_sale/static/src/app/screens/receipt_screen/receipt/order_receipt.xml:8]: uh it was just some notes :')

File: addons/point_of_sale/static/src/app/components/order_widget/order_widget.xml (hunk 15 above)
Comment (davidmonnom) [reply addons/point_of_sale/static/src/app/components/order_widget/order_widget.xml:57]: In order_receipt

**Review by caburj — APPROVED** [review]
robodoo r+

**Pull request closed without merging**